sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1506,6 +1545,52 @@ def process_bundle(self, inputs, expected_outputs):
     return result, split_results
 
 
+class ParallelBundleManager(BundleManager):
+
+  def _check_inputs_split(self, expected_outputs):
+    # We skip splitting inputs when timer is set, because operations are not
+    # triggered until we sent inputs for timers.
+    for _, pcoll_id in expected_outputs.items():
+      kind = split_buffer_id(pcoll_id)[0]
+      if kind in ['timers']:
+        return False
+
+    return True
+
+  def process_bundle(self, inputs, expected_outputs, num_workers=None):
+    num_workers = num_workers or self._num_workers
+    param_list = []
+
+    if self._check_inputs_split(expected_outputs):
+      for name, input in inputs.items():
+        for part in input.partition(num_workers):
+          param_list.append(({name : part}, expected_outputs))
+    else:
+      param_list.append((inputs, expected_outputs))
+
+    merged_result = None
+    split_result_list = []
+    with futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
+      for result, split_result in executor.map(lambda p: BundleManager(
+          self._controller, self._get_buffer, self._get_input_coder_impl,
+          self._bundle_descriptor, self._progress_frequency,
+          self._registered).process_bundle(*p), param_list):
```

Comment:
As expected_outputs never changes, you could simplify this by doing

`executor.map(lambda part: BundleManager(...).process_bundle(part, expected_outputs), partitioned_inputs)`